clang/include/clang/Basic/SyclOptReportHandler.h

```diff
@@ -0,0 +1,57 @@
+//===---------------------- SyclOptReport.h ---------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// Defines clang::SyclOptReport class.
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_BASIC_SYCLOPTREPORTHANDLER_H
+#define LLVM_CLANG_BASIC_SYCLOPTREPORTHANDLER_H
+
+#include "clang/Basic/SourceLocation.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/StringRef.h"
+
+namespace clang {
+
+class FunctionDecl;
+
+class SyclOptReportHandler {
+private:
+  struct OptReportInfo {
+    std::string KernelArgName;
+    std::string KernelArgType;
+    SourceLocation KernelArgLoc;
+
+    OptReportInfo(std::string ArgName, std::string ArgType,
+                  SourceLocation ArgLoc)
+        : KernelArgName(ArgName), KernelArgType(ArgType), KernelArgLoc(ArgLoc) {
+    }
+  };
+  llvm::DenseMap<const FunctionDecl *, SmallVector<OptReportInfo, 4>> Map;
+
+public:
+  void AddKernelArgs(const FunctionDecl *FD, std::string ArgName,
+                     std::string ArgType, SourceLocation ArgLoc) {
+    Map[FD].emplace_back(ArgName, ArgType, ArgLoc);
+  }
+  SmallVector<OptReportInfo, 4> &GetInfo(const FunctionDecl *FD) {
```

Comment:
Rather than `Has` and `Get` that do very similar things, how about: `Optional<SmallVectorImpl<OptReportInfo>&>` (Note, I can never remember whether LLVM's `Optional` allows you to wrap a reference, so if that doesn't work, then feel free to ignore the suggestion.)

This would allow callers to call `GetInfo()` and rather than getting an assertion if they forgot to call `HasOptReportInfo()`, they get `llvm::None` instead.

Reply:
That didn't work!

Reply:
Bummer! Thanks for trying!

Then I think the only other suggestion I have is to return a `SmallVectorImpl<OptReportInfo>` instead of the concrete `SmallVector<>` type, if possible. This way, callers can decide for themselves what constitutes "small" and aren't stuck agreeing that `4` is it.